DOCS-2315 Getting Started with Multistep API Tests Edits (#11307)
DataDog/documentation: content/en/getting_started/synthetics/api_test.md

---

## Overview

> API tests **proactively monitor** that your most important services are available at anytime and from anywhere. [Single API tests][1] come in five different subtypes (HTTP, SSL, DNS, TCP, ICMP) that allow you to launch requests on the **different network layers** of your systems. Multistep API tests enable you to run HTTP tests in sequence to **monitor the uptime of key journeys at the API level**.

Author: ```suggestion
API tests proactively monitor that your most important services are available at anytime and from anywhere. [Single API tests][1] come in five different subtypes (HTTP, SSL, DNS, TCP, ICMP) that allow you to launch requests on the different network layers of your systems. Multistep API tests enable you to run HTTP tests in sequence to **monitor the uptime of key journeys at the API level**.
```

## Create a single API test

> The example below demonstrates the creation of an [HTTP test][2], a subtype of [single API tests][1]. [HTTP tests][1] **monitor your API endpoints** and **alert you** when they become too slow or fail. These tests verify that your applications are responding to requests and meeting any conditions you define, such as expected response time, HTTP status code, and header or body contents.

Author: ```suggestion
The example below demonstrates the creation of an [HTTP test][2], a subtype of [single API tests][1]. [HTTP tests][1] monitor your API endpoints and alert you when they become too slow or fail. These tests verify that your applications are responding to requests and meeting any conditions you define, such as expected response time, HTTP status code, and header or body contents.
```

{{< img src="getting_started/synthetics/assertions-example-2.png" alt="Default assertions" style="width:100%;">}}

Assertions are fully customizable. To add a custom assertion, click on elements of the response preview like the headers or click **New Assertion** to define a new assertion from scratch. 

{{< img src="getting_started/synthetics/api-test-configuration-2.mp4" alt="Example API test configuration" video="true"  >}}

**Note**: You can also leverage [global][7] and [local][8] variables in your assertions.

### Select locations 

Select one or more **Managed Locations** or **[Private Locations][9]** to run your test from.

Managed locations allow you to test public-facing websites and endpoints. To test internal applications or simulate user behavior in discrete geographic regions, select one of your private locations instead.

The Shopist application is publicly available, you can consequently go ahead and pick any managed locations you want to execute your test from.

### Specify test frequency

Select the frequency at which you want your test to execute. You can leave the default frequency of 1 minute.

> **Note:** In addition to running your Synthetic test on a schedule, you can trigger them manually or directly from your [CI/CD pipelines][10]. 

Author: ```suggestion
**Note:** In addition to running your Synthetics test on a schedule, you can trigger them manually or directly from your [CI/CD pipelines][10]. 
```

The example test below demonstrates the creation of a multistep API test, which monitors the addition of an item to a cart. It contains three steps: 
- Getting a cart, 
- Getting a product, 
- And adding the product to the cart. 

If you don't know which API endpoints to create your multistep API test on, use the example endpoints below. 

To create a new multistep API test, click **New Test** > **[Multistep API test][5]**. Add a test name (such as `Add product to cart`), include tags, and select locations. 

### Get a cart

1. In **Define steps**, click **Create Your First Step**. 
2. Add a name to your step, for example: `Get a cart`.
3. Specify the HTTP method and the URL you want to query. Here, you can add: `POST` and `https://api.shopist.io/carts`. 
4. Click **Test URL**. This creates a cart item in the Shopist application's backend.
5. Leave the default assertions or modify them.
6. Optionally, define execution parameters. 

>     Select **Continue with test if this step fails** is helpful to ensure a whole endpoint collection is tested or to ensure the last cleanup step is executed regardless of previous steps' success or failure. The **Retry** step feature can come in handy in situations where you know your API endpoint may take some time before responding. 

Author: ```suggestion
    Selecting **Continue with test if this step fails** is helpful to ensure a whole endpoint collection is tested or to ensure the last cleanup step is executed, regardless of previous steps' success or failure. The **Retry** step feature is handy in situations where you know your API endpoint may take some time before responding. 
```

## Create a multistep API test

> [Multistep API tests][12] enable you to **monitor key business transactions at the API level**. Similar to [HTTP tests][2], [multistep API tests][12] can **alert you** when your endpoints become too slow or fail to meet any conditions you defined. [Multistep API test][12] also come with the ability to create variables from individual step responses and re-inject their values in subsequent steps, allowing you to chain steps together in a way that mimics the behavior of your application.

Author: ```suggestion
[Multistep API tests][12] enable you to **monitor key business transactions at the API level**. Similar to [HTTP tests][2], [multistep API tests][12] alert you when your endpoints become too slow or fail to meet any conditions you defined. 

With [multistep API tests][12], you can create variables from individual step responses and re-inject their values in subsequent steps, chaining steps together in a way that mimics the behavior of your application.
```

> The example below demonstrates the creation of a [multistep API test][12] that monitor the addition of an item to a cart. If you don't know which API endpoints to create your [multistep API test][12] on, you can use the same endpoints as in the below example. 

Author: ```suggestion
The example below demonstrates the creation of a [multistep API test][12], which monitors the addition of an item to a cart. If you don't know which API endpoints to create your [multistep API test][12] on, use the example endpoints below. 
```

> To create a new [multistep API test][12], go to **New Test** and click on **[Multistep API test][5]**. Just like for your HTTP test, add a name (e.g., `Add product to cart`), select tags, and select locations for your [Multistep API test][12]. 

Author: ```suggestion
To create a new [multistep API test][12], click **New Test** and **[Multistep API test][5]**. Just like for your HTTP test, add a name (e.g., `Add product to cart`), select tags, and select locations for your [Multistep API test][12]. 
```

> The test created below has three steps: getting a cart, getting a product, and adding the product to the cart.

Author: ```suggestion
This example test contains three steps: getting a cart, getting a product, and adding the product to the cart.
```

### Get a cart

1. In **Define steps**, click **Create Your First Step**. 

> 2. Add a name to your step, for example `Get a cart`.

Author: ```suggestion
2. Add a name to your step, for example: `Get a cart`.
```

> 3. Specify the HTTP method and the URL you want to query. Here, you can select `POST` and use `https://api.shopist.io/carts`. Click **Test URL**. This creates a cart item in the backend of the same Shopist application previously used.
> 4. Leave the default assertions or optionally modify them.
> 5. Optionally, define execution parameters. The `Continue with test if this step fails` is useful if you want to ensure a whole endpoint collection gets tested even if one of the step fails, or if you want to ensure your last cleanup step is being performed regardless of previous steps' success or failure. The retry feature can come in handy in cases where you know your API endpoint might expectedly take some time before responding. In this example, no specific execution parameter is needed.
> 6. Optionally, extract variables from the response content by clicking **Extract a variable from response content** and entering variables names and definitions. Here, the value that needs to be extracted is the cart id which is located at the very end of the value of the `location` header:
>     - Call your variable `CART_ID`.
>     - Click on the `location` header to select it.
>     - Add a regular expression in the **Parsing Regex** field such as `(?:[^\\/](?!(\\|/)))+$` to extract the id of the cart.
> 7. Click **Save Variable**.
> 8. When you're done creating this variable, click **Save Step**.

Author: ```suggestion
3. Specify the HTTP method and the URL you want to query, for example: `POST` and `https://api.shopist.io/carts`. 
4. Click **Test URL**. This creates a cart item in the Shopist application's backend.
5. Optionally, leave the default assertions or modify them.
6. Optionally, define execution parameters. Select **Continue with test if this step fails** to ensure a whole endpoint collection is tested even if one of the step fails or to ensure your last cleanup step performs regardless of previous steps' success or failure. The retry test feature is helpful in situations where you know your API endpoint may take some time before responding. In this example, no specific execution parameter is needed.
7. Optionally, extract variables from the response content. Click **Extract a variable from response content** and enter a variable name. 
8. To extract the value of the cart ID located at the end of the `location` header:
    - Name your variable as `CART_ID`.
    - In the **Response Header,** select `location`.
    - In the **Parsing Regex** field, add a regular expression such as `(?:[^\\/](?!(\\|/)))+$`.
9. Click **Save Variable**.
10. When you're done creating this test step, click **Save Step**.
```

### Get a product
   

> 1. Click **Add Another Step** to add another step. By default, you can create up to ten steps.
> 2. Add a name to your step, for example `Get a product`.
> 3. Specify the HTTP method and the URL you want to query. Here, you can select `GET` and use `https://api.shopist.io/products.json`. Click **Test URL**. This gets you a list of products available on the Shopist application.
> 4. Leave the default assertions or optionally modify them.
> 5. Optionally, define execution parameters. No specific execution parameter is needed in this step either.
> 6. Optionally, extract variables from the response content by clicking **Extract a variable from response content** and entering variables names and definitions. Here, the value that needs to be extracted is the product id which is located in the response body:
>     - Call your variable `PRODUCT_ID`.
>     - Click on the **Response Body** tab.
>     - Look for the `$oid` of any product and click it to generate a JSON Path such as `$[0].id['$oid']` for instance.
> 7. Click **Save Variable**.
> 8. When you're done creating this variable, click **Save Step**.

Author: ```suggestion
1. In **Define another step**, click** **Add Another Step**. By default, you can create up to ten steps.
2. Add a name to your step, for example: `Get a product`.
3. Specify the HTTP method and the URL you want to query, for example: `GET` and `https://api.shopist.io/products.json`. 
4. Click **Test URL**. This retrieves a list of products available in the Shopist application.
5. Optionally, leave the default assertions or modify them.
6. Optionally, define execution parameters. In this example, no specific execution parameter is needed.
7. Optionally, extract variables from the response content. Click **Extract a variable from response content** and entering a variable name.
8. To extract the value of the product ID located in the response body:
    - Name your variable as `PRODUCT_ID`.
    - Click the **Response Body** tab.
    - Click on `$oid` of any product to generate a JSON Path such as `$[0].id['$oid']`.
8. Click **Save Variable**.
9. When you're done creating this test step, click **Save Step**.
```

### Add product to cart

1. Click **Add Another Step** to add another step about adding a product into your cart.
2. Add a name to your step, for example `Add product to cart`.
3. Specify the HTTP method and the URL you want to query. Here, you can select `POST` and use `https://api.shopist.io/add_item.json`. 
4. In the **Request Body** tab, choose `application/json` body type and insert the following request body:

        ```
        {
          "cart_item": {
            "product_id": "{{ PRODUCT_ID }}",
            "amount_paid": 500,
            "quantity": 1
          },
          "cart_id": "{{ CART_ID }}"
        } 
        ```
        
5. Click **Test URL**. This adds the product you extracted in step 2 to the cart you created in step 1 and returns a checkout URL.
6. Leave the default assertions or optionally modify them.
7. Optionally, define execution parameters. No specific execution parameter is needed in this step either.
8. Optionally, extract variables from the response content. No specific variables needs to be extracted in this step.
9. When you're done creating this variable, click **Save Step**.

> 

Author: ```suggestion

{{< img src="getting_started/synthetics/defined-steps.png" alt="Created test steps" style="width:100%;" >}}
```

### Configure the test

When you create a [multistep API test][12], you can define a sequence of HTTP requests and create variables from the response data to re-inject their values in subsequent steps. 

1. In **New Test** > **[Multistep API test][5]**, add a name, select a tag, and select locations for your [Multistep API test][5].
2. In **Define steps**, click **Create Your First Step**. 

    - Add a name to your step such as `Get a cart`.
    - Specify the URL you want to query. If you don’t know what to start with, select `POST` and use `https://api.shopist.io/carts`, a test application.
    - Select **Advanced Options** to add custom request headers, authentication credentials, body content, or cookies.
    - Click **Test URL**. 
   

>    Optionally, add an assertion based on information in the **Response Body** or click inside the **Response Header**.

Reviewer: These options look out of place without being added to the indented list. Let's add to the existing list above.

   
   Optionally, define the execution parameter by enabling the test to continue if the step fails or specifying the number of times to retry the test after a specific amount of milliseconds in case of failure.
   
   Optionally, extract variables from the response content by clicking **Extract a variable from response content** and entering a variable name such as `CART_ID`. 

   To extract a variable from the response header:

    - Click inside the **Response Header** to specify which header you want to extract a value from or specify a header such as `location` in the **Header from which to extract value** field.
    - Add a regular expression in the **Parsing Regex** field such as `(?:[^\\/](?!(\\|/)))+$` to extract the location of the cart.

3. Click **Save Variable**.
4. When you're done creating this variable, click **Save Step**.
5. Click **Add Another Step** to add another step. By default, you can create up to ten steps.
    - Add a name to your step such as `Get a product`.
    - For the URL, select `GET` and use `https://api.shopist.io/products.json`.
    - Configure any **Advanced Options** and click **Test URL**.
  

>    Optionally, add an assertion based on information in the **Response Body** or click inside the **Response Header**.

Reviewer: Likewise with added these to the existing list

   
   To extract a variable from the response body:

    - Scroll through the **Response Body**, select a query language (JSON Path, XPath 1.0, or Regex) to parse the body content with, and add an expression in the **Parse** or **Path** field. 
    - To automatically generate the JSON path for the variable, click on the response body content such as `$oid:`. 
    - Optionally, you can select **Use Full Response Body**.

6. When you're done creating this variable, click **Save Step**.
7. Click **Add Another Step** to add another step about adding a product into your cart.
    - Add a name to your step such as `Add product to cart`.
    - For the URL, select `POST` and use `https://api.shopist.io/add_item.json`.
    - In the **Request Header** field of **Advanced Options** > **Request Options**, add `content-type` as the name and `application/json` as the value.
    - In the **Request Body** field, insert the following:
        ```
        {
          "cart_item": {
            "product_id": "{{ PRODUCT_ID }}",
            "amount_paid": 500,
            "quantity": 1
          },
          "cart_id": "{{ CART_ID }}"
        } 
        ```

    - Click **Test URL**.

>    Optionally, extract variables from the response content by clicking **Extract a variable from response content** and entering a variable name such as `CHECKOUT_URL`. 

Reviewer: Here, too.
